### PR TITLE
Treat the isProbablyReaderable verdict as advisory so it can't discard the parse

### PR DESCRIPTION
**Pull request by twlatl**
`ReadabilityBasic.js` posts its `isProbablyReaderable()` verdict *before* it runs the parse:

```js
if (isProbablyReaderable(document)) { postStateChanged("Available") }
else { postStateChanged("Unavailable") }

var documentClone = document.cloneNode(true);
const readabilityResult = new Readability(documentClone, __READABILITY_OPTION__).parse();
```

`ReadabilityRunner.parseHTML` throws on an `Unavailable` verdict and unsubscribes, so the `contentParsed` message that follows lands on a dead subscription and is dropped. Because the message handler decodes `contentParsed` inside a `Task.detached`, it is *always* the later of the two — the race is a certainty, not a chance.

### The pre-check is the wrong gate

`isProbablyReaderable` scores only `p, pre, article` nodes (plus `div > br` parents) holding at least `minContentLength: 140` characters, and needs a total above `minScore: 20`. It cannot see lists, tables or figures at all. Firefox uses it to decide whether to *offer* the reader button, not to decide whether parsing may proceed.

### Reproduction

https://photorumors.com/2026/09/16/canon-eos-r8-ii-camera-officially-announced/ — a short post whose body is one 79-character intro line, a `<ul>` of links, and images. Its only paragraph long enough to count is a 453-character affiliate disclaimer, scoring `sqrt(453 - 140) ~= 17.7`, just under 20 — so the verdict is `Unavailable`:

```
Readability().parse(url:)   ->  throws readerIsUnavailable
ReadabilityWebCoordinator   ->  availability: unavailable, then available
                            ->  contentParsed: 15991 chars
```

Same page, same scripts: the façade throws while the coordinator — whose availability and content are independent streams — parses it fine. The content was there the whole time.

### The change

An `Unavailable` verdict now opens a short grace window instead of failing outright. Content arriving within it wins; if none arrives, the page genuinely had nothing to extract and `readerIsUnavailable` is still the answer.

The window is load-bearing rather than cosmetic: a `Readability.parse()` that returns null stringifies to `"null"`, fails to decode into `ReadabilityResult`, and emits no event at all — so without a deadline there would be no terminal event to wait for and those pages would hang. That is the same shape as #5, from the other direction.

Outcomes funnel through a small one-shot resolver, since both halves of the conversation can fire and a `CheckedContinuation` resumed twice traps.

### Compatibility

No public API change, no new dependency. Seconds rather than `Duration`, to keep the declared macOS 11 / iOS 14 floors. `swift build` passes.

---

Found while building an RSS reader that uses this package for reader mode — thanks for maintaining it. Happy to adjust the grace period, make it configurable through `Readability.Options`, or add a regression test if you'd like one (the test target is currently empty, so I didn't want to introduce a harness unasked).

### CODE REVIEW

File: Sources/Readability/Internal/ReadabilityRunner.swift [hunk 1]
```diff
@@ -51,15 +51,18 @@ final class ReadabilityRunner {
...
                 switch event {
                 case let .contentParsed(readabilityResult):
-                    continuation.resume(returning: readabilityResult)
-                    self?.messageHandler?.subscribeEvent(nil)
+                    resolver.resolve(.success(readabilityResult))
                 case let .availabilityChanged(availability):
                     if availability == .unavailable {
-                        continuation.resume(throwing: Error.readerIsUnavailable)
-                        self?.messageHandler?.subscribeEvent(nil)
+                        resolver.failIfNoContentArrives(withinSeconds: 2,
```
Comment (Ryu0118) [inline Sources/Readability/Internal/ReadabilityRunner.swift]: **Grace window starts before the JS parse even begins, so 2s may be too tight — and the failure-latency change deserves a doc note.**

The native timer is armed when `StateChange` reaches Swift, but the script posts that verdict *before* it does any of the real work:

```js
isProbablyReaderable(document) ? t("Available") : t("Unavailable");
var a = document.cloneNode(true);
const r = new Readability(a, __READABILITY_OPTION__).parse();
postMessage({ Type: "ContentParsed", Value: JSON.stringify(r) });
```

So the 2s budget has to cover `cloneNode` + `parse()` + `JSON.stringify` + the `Task.detached` `JSONDecoder` pass in `ReadabilityMessageHandler` + the hop back to `MainActor`. The sanitized path (`ReadabilitySanitized.js`) is heavier still: `XMLSerializer.serializeToString` → DOMPurify with `WHOLE_DOCUMENT` → `DOMParser` → `parse()`. On a long page or an older device the pages this PR is trying to rescue (list/table/figure-heavy, which the pre-check already scored as "thin") could plausibly blow past 2s and land right back on a false `readerIsUnavailable`.

Two suggestions, either or both:

1. Give the window more headroom and pull it out into a named constant (e.g. `static let availabilityGracePeriod: TimeInterval = 5`) so the number is discoverable and has a comment explaining what it's meant to outlast.
2. Optionally expose it through `Readability.Options`. One wrinkle: `Options` is `Encodable` and is serialized wholesale into `__READABILITY_OPTION__`, so a Swift-only field would be shipped to the JS side too. `shouldSanitize` already does that today, so there is precedent, but a `CodingKeys` exclusion would keep the JS options object clean.

Separately, this changes observable behaviour for genuinely-unreadable pages: they now fail after the grace period instead of immediately. That's a reasonable trade, but the doc comment on `Error.readerIsUnavailable` ("the reader became unavailable during parsing") no longer tells the whole story — it now also means "no parse output arrived within the grace window". Worth updating that comment (and a line in the PR description) so callers know to expect the delay.

File: Sources/Readability/Internal/ReadabilityRunner.swift [hunk 2]
```diff
@@ -69,6 +72,39 @@ final class ReadabilityRunner {
...
+    func resolve(_ outcome: Result<ReadabilityResult, Swift.Error>) {
+        guard let continuation else { return }
+        self.continuation = nil
+        graceTask?.cancel()
+        graceTask = nil
+        onFinish()
+        continuation.resume(with: outcome)
+    }
+
+    func failIfNoContentArrives(withinSeconds seconds: Double, with error: Swift.Error) {
```
Comment (Ryu0118) [inline Sources/Readability/Internal/ReadabilityRunner.swift]: **Design alternative worth considering: make the null-parse case emit a terminal event instead of relying on a timer.**

As the PR description notes, the only reason a deadline is load-bearing is that `Readability.parse()` returning `null` stringifies to `"null"`, fails `JSONDecoder().decode(ReadabilityResult.self, ...)` in `ReadabilityMessageHandler.userContentController(_:didReceive:)` (lines 52–66), and silently emits nothing. That same silence also means the timer only helps in one direction: an `.available` verdict followed by a `null` parse still hangs forever after this change.

A timer-free shape that fixes both directions deterministically:

- In `ReadabilityMessageHandler`, when the `.contentParsed` decode fails, emit a new terminal case (e.g. `Event.contentParseFailed`) instead of dropping the message.
- In `ReadabilityRunner`, treat `.contentParseFailed` as `readerIsUnavailable` and ignore `.availabilityChanged` entirely for the purposes of resolving the continuation (it becomes purely advisory, which matches the PR's framing).

Two things to be aware of if you go that way:

- `Event` is `package`-scoped so this isn't a public API break, but `ReadabilityWebCoordinator.createReadableWebViewConfiguration()` (`Sources/ReadabilityUI/ReadabilityWebCoordinator.swift:75–83`) has an exhaustive `switch` with no `default`, so it will need a `case .contentParseFailed: break` (or a yield on the availability stream).
- The JS side does not wrap `parse()` in `try/catch`. Readability throws when `maxElemsToParse` is exceeded, and in that case *no* `ContentParsed` message is posted at all, so a decode-failure event alone still leaves that path hanging under `.available`. A `try { ... } catch { postMessage({Type: "ContentParsed", Value: "null"}) }` in `ReadabilityBasic.js` / `ReadabilitySanitized.js` would close that gap and let the grace timer go away completely.

Not a blocker: the current approach is already a clear improvement over the deterministic drop. Just flagging that the "some pages still hang" asymmetry you called out has a fairly contained fix if you'd like to take it in this PR or a follow-up.

File: Sources/Readability/Internal/ReadabilityRunner.swift [hunk 3]
```diff
@@ -69,6 +72,39 @@ final class ReadabilityRunner {
     }
 }
 
+@MainActor
+private final class ParseResolver {
```
Comment (Ryu0118) [inline Sources/Readability/Internal/ReadabilityRunner.swift]: **`ParseResolver` is a good candidate for the regression test you offered — it doesn't need a `WKWebView`.**

It's a plain state machine over a `CheckedContinuation` and a `Task`, so it can be driven directly from `withCheckedThrowingContinuation` in a test. The `ReadabilityTests` target already exists (`dependencies: ["Readability"]`, currently an empty file), so `@testable import Readability` would work as-is; the only change needed here is `private` → `internal` so the test target can see the type.

Three cases would pin down the behaviour this PR depends on:

1. `failIfNoContentArrives` followed by `resolve(.success(...))` before the deadline resolves with the content and cancels the grace task (no later throw).
2. `failIfNoContentArrives` with no subsequent `resolve` resolves with `readerIsUnavailable` after the deadline (a small `withinSeconds:` value keeps it fast).
3. Calling `resolve` twice — or `resolve` after the deadline has already fired — is a no-op, which is the double-resume trap the resolver exists to prevent.

Also worth asserting that `onFinish` is invoked exactly once in each of the above, since that's what unsubscribes the message handler.

**Review by Ryu0118 — APPROVED** [review]
Verified:
- `swift build` passes on the PR branch (Swift 6 language mode, all targets)
- `swift test` passes (ParseResolver behavior)
- Also ran the ParseResolver test suite on an iOS simulator (via the Example app's local package reference) — passes there too
- Reproduced the underlying race the PR describes and confirmed the fix resolves it

Left a few non-blocking suggestions on the deadline window sizing, an alternative timer-free design, and test coverage for `ParseResolver` — none of them block merging. This is a clear improvement over the current behavior of silently dropping a valid parse. Thanks for the thorough writeup and repro.

**Comment by Ryu0118** [comment]
@twlatl  Thank you for your contribution! These changes have been released in v0.4.0:
https://github.com/Ryu0118/swift-readability/releases/tag/0.4.0
